src/backend/distributed/transaction/transaction_management.c

```diff
@@ -396,6 +396,17 @@ CoordinatedSubTransactionCallback(SubXactEvent event, SubTransactionId subId,
 {
 	switch (event)
 	{
+		/*
+		 * Our subtransaction stack should be consistent with postgres' internal
+		 * transaction stack. In case of subxact begin, postgres calls our
+		 * callback after it has pushed the transaction into stack, so we have to
+		 * do the same even if worker commands fail, so we PushSubXact() first.
+		 * In case of subxact commit, callback is called before pushing subxact to
+		 * the postgres transaction stack, so we call PopSubXact() after making sure
+		 * worker commands didn't fail. For subxact abort, we don't care about the
```

Comment:
Maybe add: "Otherwise, Postgres would roll back that would cause us to call PopSubXact again".

The last sentence can probably be removed.